### PR TITLE
[rocdl] Add e2e matmul test for cdna3 matrix core

### PR DESCRIPTION
**Pull request by antiagainst**
This commit adds support for matmul correctness
test over CodeGen paths that target cnda3 mfma.
Right now we only cover (M, K) x (K, N) -> (M, N)
matmul for v_mfma_f32_16x16x16_f16.

### CODE REVIEW

**Comment by antiagainst** [comment]
Depends on https://github.com/openxla/iree/pull/16526

File: build_tools/cmake/iree_e2e_matmul_test.cmake [hunk 1]
```diff
@@ -216,12 +216,6 @@ function(iree_single_backend_e2e_matmul_test)
   if(DEFINED _RULE_DRIVER)
     string(TOUPPER ${_RULE_DRIVER} _UPPERCASE_DRIVER)
     string(REPLACE "-" "_" _NORMALIZED_DRIVER ${_UPPERCASE_DRIVER})
-    string(TOUPPER "${IREE_EXTERNAL_HAL_DRIVERS}" _UPPERCASE_EXTERNAL_DRIVERS)
```
Comment (antiagainst) [inline build_tools/cmake/iree_e2e_matmul_test.cmake:219]: @ScottTodd as you commented in https://github.com/openxla/iree/pull/16466#discussion_r1494788870, this is an issue for external targets that are not explicitly turned on. So need to remove it.

File: tests/e2e/matmul/CMakeLists.txt [hunk 2]
```diff
@@ -976,6 +976,32 @@ iree_generated_e2e_matmul_test(
     "noriscv"
 )
 
+iree_generated_e2e_matmul_test(
+  NAME
+    e2e_matmul_rocm_f16_large_cdna3_matrixcore
```
Comment (ScottTodd) [inline tests/e2e/matmul/CMakeLists.txt:981]: Is this running on CI anywhere? Or just for local use at the moment? I do want to set up `test_gpu_nvidia` and `test_gpu_amd` jobs once we have enough test runners.

We'll basically have a matrix around this to run tests on different machines: https://github.com/openxla/iree/blob/5d8907e82fc1eb741a4d4d27f5cae865323fd1d7/.github/workflows/ci.yml#L268-L388

(sort of like the `cross_compile_and_test` matrix, ... I need to take a closer look at that to see if it should exist separately or fold into that too)

File: build_tools/cmake/iree_e2e_matmul_test.cmake (hunk 1 above)
Comment (ScottTodd) [reply build_tools/cmake/iree_e2e_matmul_test.cmake:219]: I'm not sure I follow ... if the check below (lines 219-220) works then I don't see why this wouldn't also work?

Fine to remove the overly aggressive `SEND_ERROR` regardless though. Might want to do that in `iree_check_test.cmake` also for consistency.

Comment (antiagainst) [reply build_tools/cmake/iree_e2e_matmul_test.cmake:219]: Sorry for the confusion. In #16466, the whole HIP subdirectory is gated by a big switch as experimental external HAL driver, so it works. Here this specific test does not have that kind of big switch, so it doesn't work.

File: tests/e2e/matmul/CMakeLists.txt (hunk 2 above)
Comment (antiagainst) [reply tests/e2e/matmul/CMakeLists.txt:981]: Local for now to make it easy to debug matmul correctness issues. But yes!!! I'd certainly want continous CI testing for it.

File: build_tools/cmake/iree_e2e_matmul_test.cmake (hunk 1 above)
Comment (antiagainst) [reply build_tools/cmake/iree_e2e_matmul_test.cmake:219]: Okay dropped the `SEND_ERROR` in `iree_check_test.cmake` too for consistency.

**Review by ScottTodd — APPROVED** [review]

**Review by qedawkins — APPROVED** [review]
(merging because I need this and I think others are currently afk)

**Pull request merged**